File: Javascript/35.search-insert-position.js
/*
 * @lc app=leetcode id=35 lang=javascript
 *
 * [35] Search Insert Position
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var searchInsert = function(nums, target) {
    let start = 0;
    let end = nums.length - 1;

    while (start <= end) {
        if (nums[start] === target) {
            return start;
        }

        if (nums[end] === target) {
            return end;
        }

        start++;
        end--;
    }

    let tobe = 0;
    for (i = nums.length - 1; i >= 0; i--) {
        if (nums[i] < target && tobe < i + 1) {
            tobe = i + 1;
        }
    }

    return tobe;
};
// @lc code=end

console.log(searchInsert([1,3,5,6], 7))

File: Ruby/35.search-insert-position.rb
#
# @lc app=leetcode id=35 lang=ruby
#
# [35] Search Insert Position
#

# @lc code=start
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer}
def search_insert(nums, target)
    return 0 if nums.nil?
    left = 0
    right = nums.size - 1
    while left <= right do
        mid = (right - left) / 2 + left
        if nums[mid] == target
            return mid
        elsif target > nums[mid]
            left = mid + 1
        else
            right = mid - 1
        end
    end
    left
end
# @lc code=end

search_insert([1, 3, 5, 6], 2)
